source/pdal/pdalc_pipeline.cpp

```diff
@@ -103,208 +104,228 @@ extern "C"
+                return 0;
             }
-
         }
-
-        return result;
+        return 0;
     }
 
     size_t PDALGetPipelineMetadata(PDALPipelinePtr pipeline, char *metadata, size_t size)
     {
-        size_t result = 0;
-
         if (pipeline && metadata && size > 0)
         {
             Pipeline *ptr = reinterpret_cast<Pipeline *>(pipeline);
-            pdal::PipelineExecutor *executor = ptr->get();
-            metadata[0] =  '\0';
-            metadata[size - 1] = '\0';
 
-            if (executor)
+            try
+            {
+                if (! ptr->m_executed)
+                    throw pdal_error("Pipeline has not been executed!");
+
+                std::stringstream strm;
+                MetadataNode root = ptr->manager->getMetadata().clone("metadata");
+                pdal::Utils::toJSON(root, strm);
+                std::string out = strm.str();
+                if (out.length() > size - 1) out.resize(size - 1);
+                std::strncpy(metadata, out.c_str(), size);
+                return std::min(out.length() + 1, size);
+            }
+            catch (const std::exception &e)
             {
-                try
-                {
-                    std::string s = executor->getMetadata();
-                    std::strncpy(metadata, s.c_str(), size);
-                    result = std::min(s.length(), size);
-                }
-                catch (const std::exception &e)
-                {
-                    printf("Found error while retrieving pipeline's metadata: %s\n", e.what());
-                }
+                printf("Found error while retrieving pipeline's metadata: %s\n", e.what());
+                return 0;
             }
         }
-
-        return result;
+        return 0;
     }
 
     size_t PDALGetPipelineSchema(PDALPipelinePtr pipeline, char *schema, size_t size)
     {
-        size_t result = 0;
-
         if (pipeline && schema && size > 0)
         {
             Pipeline *ptr = reinterpret_cast<Pipeline *>(pipeline);
-            pdal::PipelineExecutor *executor = ptr->get();
-            schema[0] =  '\0';
-            schema[size - 1] = '\0';
 
-            if (executor)
+            try
             {
-                try
-                {
-                    std::string s = executor->getSchema();
-                    std::strncpy(schema, s.c_str(), size);
-                    result = std::min(s.length(), size);
-                }
-                catch (const std::exception &e)
-                {
-                    printf("Found error while retrieving pipeline's schema: %s\n", e.what());
-                }
+                std::stringstream strm;
+                MetadataNode meta = ptr->manager->pointTable().layout()->toMetadata();
+                MetadataNode root = meta.clone("schema");
+                pdal::Utils::toJSON(root, strm);
+                std::string out = strm.str();
+                if (out.length() > size - 1) out.resize(size - 1);
```

Comment:
Could do:
```
memset(schema, 0, size);
...
/// This will fill with NULLs as necessary, so you'll have at most size-1 non-null characters.
out.resize(size - 1);
// Copy the properly sized string to output.
strcpy(schema, out.c_str());
// This is inefficient but simple and I'd vote simplicity over efficiency here.
return strlen(schema);  // Or strlen(schema) + 1 -- not sure if you're returning the string length with or without the null terminator.
```

Reply:
hmm

- no need to initialize schema etc since strncpy will pad with zeros if the string length is less than the buffer size.
- we can't just resize out with the test since in the majority of cases where the contents are less than the buffer it will error out

Reply:
suggestion of strlen accepted, other suggestions not accepted